### PR TITLE
Change how progress notifications are dispatched

### LINKED ISSUE #4623: SyncSession.addProgressNotification does not work
## Goals

Get notifications for sync upload and download progress.

## Expected Results

Get sync progress notifications.

## Actual Results

Sync progress notifications are never delivered.

## Steps to Reproduce

What are steps we can follow to reproduce this issue?

## Code Sample

```
import UIKit
import RealmSwift

class O: Object {

    dynamic var data = Data()

}

@UIApplicationMain
class AppDelegate: UIResponder, UIApplicationDelegate {

    var window: UIWindow?
    var token: NotificationToken?
    
    var realm: Realm!
    var downloadToken: SyncSession.ProgressNotificationToken?
    var uploadToken: SyncSession.ProgressNotificationToken?
    
    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplicationLaunchOptionsKey: Any]?) -> Bool {
        // Override point for customization after application launch.
        
        let serverURL = URL(string: "http://localhost:9080")!
        
        SyncUser.logIn(with: .usernamePassword(username: "username", password: "password"), server: serverURL) {
            assert($1 == nil)
            self.loginCompleted(user: $0!)
        }
        
        return true
    }
    
    func loginCompleted(user: SyncUser) {
        let realmURL = URL(string: "realm://localhost:9080/~/10k_data_multiple_writes")!
        var configuration = Realm.Configuration.defaultConfiguration
        configuration.syncConfiguration = SyncConfiguration(user: user, realmURL: realmURL)
        realm = try! Realm(configuration: configuration)
        
        let session = user.session(for: realmURL)!
        downloadToken = session.addProgressNotification(for: .download, mode: .reportIndefinitely) {
            print("download progress: \($0)") // This is never called.
        }
        uploadToken = session.addProgressNotification(for: .upload, mode: .reportIndefinitely) {
            print("upload progress: \($0)") // This is never called.
        }
        
        for _ in 0..<10_000 {
            try! realm.write {
                let o = O()
                o.data = "this is a test with some data...".data(using: .utf8)!
                realm.add(o)
            }
        }
    }
    
}
```

## Version of Realm and Tooling

Realm version: 2.4.2 / Realm Mobile Platform 1.0.0-2

Xcode version: 8.2.1

### PR DESCRIPTION
**Pull request by austinzheng**
Based on one possible interpretation of https://github.com/realm/realm-cocoa/issues/4626. See that ticket for discussion and rationale.

Fixes #4623.

### CODE REVIEW

**Comment by austinzheng** [comment]
Adding reviewers to open up discussion. (We might not want to go with this approach.)

File: Realm/RLMSyncSession.h [hunk 1]
```diff
@@ -150,8 +152,20 @@ NS_ASSUME_NONNULL_BEGIN
  */
 - (nullable RLMProgressNotificationToken *)addProgressNotificationForDirection:(RLMSyncProgressDirection)direction
                                                                           mode:(RLMSyncProgress)mode
+                                                           dispatchToMainQueue:(BOOL)dispatchToMainQueue
```
Comment (bdash) [inline Realm/RLMSyncSession.h]: This is really gross.

Comment (austinzheng) [inline Realm/RLMSyncSession.h]: Would it be preferable to just have the user explicitly specify a queue?

Comment (bdash) [inline Realm/RLMSyncSession.h]: What do we gain from the user explicitly specifying a queue? Why is that preferable to having them dispatch to whichever queue they care about from within the callback?

Comment (austinzheng) [inline Realm/RLMSyncSession.h]: If we're okay with telling users they must explicitly dispatch onto whatever queue they want, sure. (We might also not want users to 'accidentally' run expensive work in their callbacks without dispatching, since IIRC the callback is run on the sync client's thread.)

Comment (bdash) [inline Realm/RLMSyncSession.h]: But after the rest of this PR it's _not_ run on the sync client thread, it's run on a separate notification queue. The only thing they'll delay by doing expensive work is further progress callbacks.

Comment (austinzheng) [inline Realm/RLMSyncSession.h]: Oh, I see what you mean now. In that case, I agree. We specify that there's a notification queue, and if they want the work to be done elsewhere they can dispatch it to whatever queue they want. Is that satisfactory?

Comment (bdash) [inline Realm/RLMSyncSession.h]: That would be my inclination.

File: CHANGELOG.md [hunk 2]
```diff
@@ -23,6 +23,8 @@ x.x.x Release notes (yyyy-MM-dd)
   non-nil value when passed in an invalid URL.
 * `SyncSession.Progress.fractionTransferred` now returns 1 if there are no
   transferrable bytes.
+* Change the way sync progress notifications are dispatched to fix broken
```
Comment (jpsim) [inline CHANGELOG.md]: This doesn't announce that they're no longer dispatched on the same thread on which they were registered, which is technically a breaking change, but we've always allowed ourselves to make breaking changes to already broken implementations, so that's fine.

However, I'd reword this:

> * Fix sync progress notifications registered on background threads by always dispatching on a dedicated background queue.

Comment (austinzheng) [inline CHANGELOG.md]: Yeah, good call. There was a bit more about this in the "Enhancements" section originally, but I deleted that part. Thanks for pointing out the omission.

> but we've always allowed ourselves to make breaking changes to already broken implementations

That was my rationale as well.

**Comment by jpsim** [comment]
Fixes #4623.

**Comment by austinzheng** [comment]
One of the builds failed because the compiler segfaulted, the other three failed because of various CI issues.
